File: test-suite/src/test/java/io/micronaut/security/oauth2/docs/managementendpoints/SensitiveEndpointRuleReplacement.java
package io.micronaut.security.oauth2.docs.managementendpoints;

//tag::imports[]
import io.micronaut.context.annotation.Requires;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.context.annotation.Replaces;
import io.micronaut.http.HttpRequest;
import io.micronaut.inject.ExecutableMethod;
import io.micronaut.management.endpoint.EndpointSensitivityProcessor;
import io.micronaut.security.authentication.Authentication;
import io.micronaut.security.rules.SecurityRuleResult;
import io.micronaut.security.rules.SensitiveEndpointRule;
import io.micronaut.security.token.RolesFinder;
import jakarta.inject.Singleton;
import java.util.Collections;
//end::imports[]

@Requires(property = "spec.name", value = "LoggersSpec")
//tag::clazz[]
@Replaces(SensitiveEndpointRule.class)
@Singleton
public class SensitiveEndpointRuleReplacement extends SensitiveEndpointRule {
    private final RolesFinder rolesFinder;

    public SensitiveEndpointRuleReplacement(EndpointSensitivityProcessor endpointSensitivityProcessor,
                                            RolesFinder rolesFinder) {
        super(endpointSensitivityProcessor);
        this.rolesFinder = rolesFinder;
    }

    @Override
    @NonNull
    protected SecurityRuleResult checkSensitiveAuthenticated(@NonNull HttpRequest<?> request,
                                                             @NonNull Authentication authentication,
                                                             @NonNull ExecutableMethod<?, ?> method) {
        return rolesFinder.hasAnyRequiredRoles(Collections.singletonList("ROLE_SYSTEM"), authentication.getRoles())
                    ? SecurityRuleResult.ALLOWED : SecurityRuleResult.REJECTED;
    }
}
//end::clazz[]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/security/oauth2/docs/managementendpoints/SensitiveEndpointRuleReplacement.kt
package io.micronaut.security.oauth2.docs.managementendpoints

//tag::imports[]
import io.micronaut.context.annotation.Replaces
import io.micronaut.context.annotation.Requires
import io.micronaut.core.annotation.NonNull
import io.micronaut.http.HttpRequest
import io.micronaut.inject.ExecutableMethod
import io.micronaut.management.endpoint.EndpointSensitivityProcessor
import io.micronaut.security.rules.SecurityRuleResult
import io.micronaut.security.rules.SensitiveEndpointRule
import io.micronaut.security.token.MapClaims
import io.micronaut.security.token.RolesFinder
import jakarta.inject.Singleton
import org.reactivestreams.Publisher
import reactor.core.publisher.Mono

//end::imports[]

@Requires(property = "spec.name", value = "LoggersSpec")
//tag::clazz[]
@Replaces(SensitiveEndpointRule::class)
@Singleton
class SensitiveEndpointRuleReplacement(endpointSensitivityProcessor: EndpointSensitivityProcessor,
                                       private val rolesFinder: RolesFinder) : SensitiveEndpointRule(endpointSensitivityProcessor) {
    @NonNull
    override fun checkSensitiveAuthenticated(@NonNull request: HttpRequest<*>,
                                             @NonNull claims: Map<String, Any>,
                                             @NonNull method: ExecutableMethod<*, *>): Publisher<SecurityRuleResult> {
        return if (rolesFinder.hasAnyRequiredRoles(listOf("ROLE_SYSTEM"), MapClaims(claims))) {
            Mono.just(SecurityRuleResult.ALLOWED)
        } else {
            Mono.just(SecurityRuleResult.REJECTED)
        }
    }
}
//end::clazz[]
